org/mozilla/jss/util/jssutil.c

```diff
@@ -750,6 +752,106 @@ void JSS_DerefJString(JNIEnv *env, jstring str, const char *ref) {
     }
 }
 
+/************************************************************************
+** JSS_ExceptionToSECStatus
+**
+** When the JNI has thrown a known exception, convert this to a SECStatus
+** code and set the appropriate PRErrorCode.
+**
+** See jssutil.h for a list of supported exceptions.
+**
+*/
+SECStatus JSS_ExceptionToSECStatus(JNIEnv *env) {
+    jclass clazz;
+
+    jthrowable except = (*env)->ExceptionOccurred(env);
+    if (except == NULL) {
+        // No exception occurred; exit with success.
+        PORT_SetError(0);
+        return SECSuccess;
+    }
+
+    // Now we have to handle get the various cases. Each case involves looking
+    // up a class by name and comparing it via IsInstanceOf(...). We ignore
+    // failures which may occur on lookup.
+
+    // CERTIFICATE_ENCODING_EXCEPTION <-> SEC_ERROR_CERT_NOT_VALID
+    clazz = (*env)->FindClass(env, CERTIFICATE_ENCODING_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_CERT_NOT_VALID);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_EXPIRED_EXCEPTION <-> SEC_ERROR_EXPIRED_CERTIFICATE
+    clazz = (*env)->FindClass(env, CERTIFICATE_EXPIRED_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_EXPIRED_CERTIFICATE);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_NOT_YET_VALID_EXCEPTION <-> SEC_ERROR_CERT_NOT_VALID
+    clazz = (*env)->FindClass(env, CERTIFICATE_NOT_YET_VALID_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_CERT_NOT_VALID);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_PARSING_EXCEPTION <-> SEC_ERROR_BAD_DER
+    clazz = (*env)->FindClass(env, CERTIFICATE_PARSING_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_BAD_DER);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_REVOKED_EXCEPTION <-> SEC_ERROR_REVOKED_CERTIFICATE
+    clazz = (*env)->FindClass(env, CERTIFICATE_REVOKED_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_REVOKED_CERTIFICATE);
+        return SECFailure;
+    }
+
+    return SECFailure;
```

Comment:
Should we call PORT_SetError() with a generic error code in case the exception doesn't match anything?

Reply:
I've thought about that, but I don't know a good generic error code. I suppose you could throw `SEC_ERROR_BASE`, but I'm not sure what'd happen if you did.

The remaining candidates I see are:

 - `SEC_ERROR_LIBRARY_FAILURE`
 - `SEC_ERROR_BAD_DATA`
 - `SEC_ERROR_INVALID_ARGS`

In most cases, outside of what gets eventually shown to the user, I don't think NSS will interpret the error code...

Reply:
OK, I guess we probably should just leave it for now, but could you add a comment in the code to record this issue so maybe someone can fix the it properly in the future? I'll ACK this PR. Thanks!

Reply:
I realized that NSPR might have the answer. My suggestion would be `PR_UNKNOWN_ERROR` actually.